File: app/eta_list_view_controller.rb
class ETAListViewController < UIViewController
  include ControllerConstraintHelper

  def init_with_stop_name(name)
    @stop_name = name
    @etas = []

    @empty_reuse = 'PlainCell'
    @reuse = 'ETACell'

    init
  end

  def loadView
    @layout = ETAListLayout.new
    self.view = @layout.view

    @table = @layout.table
    @table.delegate = @table.dataSource = self
    self.automaticallyAdjustsScrollViewInsets = false
    navigationItem.setTitle(@stop_name)

    table_vc = UITableViewController.alloc.init
    table_vc.tableView = @table

    @refresh = UIRefreshControl.alloc.init
    @refresh.addTarget(self, action: 'refresh:', forControlEvents: UIControlEventValueChanged)
    table_vc.refreshControl = @refresh

    @first_load = true
    refresh(nil)
    Dispatch::Queue.main.async { @layout.update_constraint(self) }
  end

  def refresh(sender)
    UIApplication.sharedApplication.networkActivityIndicatorVisible = true

    Dispatch::Queue.concurrent.async do
      @etas = APIClient.get_etas(@stop_name)
      if @etas.nil?
        alert = UIAlertController.alertControllerWithTitle('Error',
          message: 'Unable to load arrivals, please check your internet connection.',
          preferredStyle: UIAlertControllerStyleAlert)
        alert.addAction(UIAlertAction.actionWithTitle('Okay', style: UIAlertActionStyleCancel, handler: nil))
        presentViewController(alert, animated: true, completion: nil)
        @etas = []
      end

      Dispatch::Queue.main.async do
        @first_load = false
        UIApplication.sharedApplication.networkActivityIndicatorVisible = false
        @refresh.endRefreshing
        @table.reloadData
      end
    end
  end

  def numberOfSectionsInTableView(table)
    1
  end

  def tableView(table, numberOfRowsInSection: section)
    return 0 if @first_load
    [1, @etas.count].max
  end

  def tableView(table, cellForRowAtIndexPath: path)
    if @etas.count == 0
      cell = table.dequeueReusableCellWithIdentifier(@empty_reuse)
      cell = UITableViewCell.alloc.initWithStyle(UITableViewCellStyleDefault, reuseIdentifier: @empty_reuse) if cell.nil?
      cell.textLabel.text = 'No upcoming arrivals'
    else
      cell = table.dequeueReusableCellWithIdentifier(@reuse)
      cell = ETACellView.alloc.initWithStyle(UITableViewCellStyleDefault, reuseIdentifier: @reuse) if cell.nil?
      cell.update(@etas[path.row])
    end

    cell
  end
end

File: Rail Line/ETAListViewController.swift
import UIKit
import IoniconsKit

protocol SettingsChangedDelegate: AnyObject {
    func settingsChanged()
}

class ETAListViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    private let station: Station
    private weak var delegate: SettingsChangedDelegate?
    private var notification: NSObjectProtocol?
    private let tableView = UITableView()
    private let refreshControl = UIRefreshControl()
    private var firstLoad = true
    private var etas: [ETA] = []

    private static let emptyIdentifier = "ETAListViewControllerCellEmpty"
    private static let reuseIdentifier = "ETAListViewControllerCell"

    init(station: Station, delegate: SettingsChangedDelegate) {
        self.station = station
        self.delegate = delegate
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = station.name

        view.backgroundColor = UIColor.systemBackground
        view.addSubview(tableView)
        tableView.snp.makeConstraints { (make) in
            make.top.equalTo(self.view.safeAreaLayoutGuide.snp.top)
            make.bottom.equalTo(self.additionalSafeAreaInsets.bottom)
            make.width.equalTo(self.view.snp.width)
        }

        tableView.dataSource = self
        tableView.delegate = self
        tableView.refreshControl = refreshControl
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: ETAListViewController.emptyIdentifier)
        tableView.register(ETACell.classForCoder(), forCellReuseIdentifier: ETAListViewController.reuseIdentifier)

        refreshControl.addTarget(self, action: #selector(refresh(_:)), for: .valueChanged)
        refreshControl.tintColor = Colors.cyan

        navigationController?.navigationBar.prefersLargeTitles = false
        navigationItem.largeTitleDisplayMode = .never

        updateStarButton()
        refresh()
    }

    override func viewDidAppear(_ animated: Bool) {
        notification = NotificationCenter.default.addObserver(forName:
            UIApplication.willEnterForegroundNotification, object: nil, queue: .main) { _ in self.refresh() }
    }

    override func viewWillDisappear(_ animated: Bool) {
        if let notification = notification {
            NotificationCenter.default.removeObserver(notification)
        }
        notification = nil
    }

    @objc func refresh(_ sender: Any) {
        refresh()
    }

    @objc func toggle(_ sender: Any) {
        toggle()
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if firstLoad { return 0 }
        return [1, etas.count].max()!
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if etas.isEmpty {
            if let cell = tableView.dequeueReusableCell(withIdentifier: ETAListViewController.emptyIdentifier) {
                cell.textLabel?.text = "No upcoming arrivals"
                cell.selectionStyle = .none
                return cell
            }
        } else {
            if let cell = tableView.dequeueReusableCell(withIdentifier: ETAListViewController.reuseIdentifier) as? ETACell {
                cell.update(eta: etas[indexPath.row])
                return cell
            }
        }
        fatalError()
    }

    private func updateStarButton() {
        let icon = Settings.isFavorited(station.id) ? Ionicons.iosStar : Ionicons.iosStarOutline
        let sizeDim = 25
        let size = CGSize(width: sizeDim, height: sizeDim)
        let image = UIImage.ionicon(with: icon, textColor: Colors.cyan, size: size)
        let highlightImage = UIImage.ionicon(with: icon, textColor: Colors.fadedCyan, size: size)

        let button = UIButton(type: .custom)
        button.setImage(image, for: .normal)
        button.setImage(highlightImage, for: .highlighted)
        button.addTarget(self, action: #selector(toggle(_:)), for: .touchUpInside)

        let barButton = UIBarButtonItem(customView: button)
        navigationItem.setRightBarButton(barButton, animated: true)
    }

    private func toggle() {
        Settings.toggle(station.id)
        updateStarButton()
        delegate?.settingsChanged()
    }

    private func refresh() {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }

            self.etas = APIClient.getETAs(stationId: self.station.id)

            DispatchQueue.main.async { [weak self] in
                self?.firstLoad = false
                self?.tableView.reloadData()
                self?.refreshControl.endRefreshing()
            }
        }
    }
}
